lib/spack/llnl/util/filesystem.py

```diff
@@ -1712,66 +1715,140 @@ def find(root, files, recursive=True):
+    # exists but is not a dir, we assume the user would want to
+    # know; likewise if it exists but we do not have permission to
+    # access it.
+    try:
+        stat_root = os.stat(root)
+    except OSError as e:
+        if e.errno == errno.EEXIST:
+            return []
+        else:
+            raise
+    if not stat.S_ISDIR(stat_root.st_mode):
+        raise ValueError(f"{root} is not a directory")
 
+    if max_depth is None:
+        max_depth = sys.maxsize
 
-@system_path_filter
-def _find_non_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict as os.list_dir
-    # can return files in any order (does not preserve stability)
-    found_files = collections.defaultdict(list)
+    if isinstance(globs, str):
+        globs = [globs]
+    # Apply normcase to regular expressions and to the filenames:
+    # this respects case-sensitivity semantics of different OSes
+    # (e.g. file search is typically case-insensitive on Windows)
+    regexes = [re.compile(fnmatch.translate(os.path.normcase(x))) for x in globs]
 
-    # Make the path absolute to have absolute path returned
+    # Note later calls to os.scandir etc. return abspaths if the
+    # input is absolute, see https://docs.python.org/3/library/os.html#os.DirEntry.path
     root = os.path.abspath(root)
 
-    for search_file in search_files:
-        matches = glob.glob(os.path.join(root, search_file))
-        matches = [os.path.join(root, x) for x in matches]
-        found_files[search_file].extend(matches)
+    found_files = collections.defaultdict(list)
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+    def _dir_id(stat_info):
+        # Note: on windows, st_ino is the file index and st_dev
+        # is the volume serial number. See
+        # https://github.com/python/cpython/blob/3.9/Python/fileutils.c
+        return (stat_info.st_ino, stat_info.st_dev)
+
+    def _log_file_access_issue(e):
+        errno_name = errno.errorcode.get(e.errno, "UNKNOWN")
+        tty.debug(f"find must skip {dir_entry.path}: {errno_name} {str(e)}")
+
+    visited_dirs = set([_dir_id(stat_root)])
+
+    # Each queue item stores the depth and path
+    # BFS traversal is important to ensure repeatable results given
+    # that we allow following symlinks to sibling dirs that are not
+    # as deep: we return the shortest path to a given file
+    dir_queue = collections.deque([(0, root)])
+    while dir_queue:
+        depth, next_dir = dir_queue.pop()
+        try:
+            dir_iter = os.scandir(next_dir)
+        except OSError:
+            # Most commonly, this would be a permissions issue, for
+            # example if we are scanning an external directory like /usr
+            continue
 
-    return answer
+        with dir_iter:
+            for dir_entry in dir_iter:
+                try:
+                    it_is_a_dir = dir_entry.is_dir(follow_symlinks=True)
```

Comment:
Regarding https://github.com/spack/spack/pull/41945#issuecomment-2446556730

`if entry.is_dir(follow_symlinks=True)` needs to be in a try/except, because it can encounter ELOOP (and in fact does as part of our checks).

That being said, I think I have honored all of the considerations in your comments.